### ISSUE TITLE
Node Periodically Cleanup

### ISSUE THREAD

**Issue by zhouhaibing089** [kubernetes/kubernetes#22970]
I felt that there should be something which handles removing stopped container and unused images. If there is not, the disk space will take up full as the time goes by.

I see there is a `imageManager` inside the package `kubelet`, which seems like a GC for images when disk usage is high, however, for the stopped/killed container, it seems there is no such cleanup.

When I setup a vagrant kubernetes cluster with only one schedulable node, the node will no longer be available when I create pod and delete again and again.

**Comment by AdoHe** [comment]
@zhouhaibing089 `ContainerGC` should be what you want. refer here for more detail [https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L441](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L441)

**Comment by zhouhaibing089** [comment]
@AdoHe I do meet a case that the node become read-only, But anyway, I am going to do some study here.

**Comment by yujuhong** [comment]
@zhouhaibing089, for a complete documentation of image/container garbage collection, see http://kubernetes.io/docs/admin/garbage-collection/
Kubelet does remove exited containers with a few pre-defined thresholds (e.g., max global #containers, max containers per pod, etc).  We recently just added a new feature (#8347) to aggressively delete containers associated with deleted pods. Let us know if you have any suggestion. We plan to improve the disk management for the next release.

**Comment by zhouhaibing089** [comment]
@yujuhong I am new at this area, so the following consideration may be a little immature..

1.  the threshold to delete image defaults to be 80%-90%, how about do this cleanup periodically? I am thinking that unused images on the node is not useful at all, the only situation it may help is that image could be reused when another scheduled pod uses it, too.
2.  I feel the documentation on container gc is a little bit hard to understand. 

> `minimum-container-ttl-duration`, minimum age for a finished container before it is garbage collected. Default is 1 minute.

If I understand correctly, the container would stay at lease 1 minute before it is being garbage collected

> `maximum-dead-containers-per-container`, maximum number of old instances to retain per container. Default is 2.

I am confused here, it would be great help if you could help to explain it more here.

> `maximum-dead-containers`, maximum number of old instances of containers to retain globally. Default is 100.

Again, IIUC, there would be at most 100 stopped container on that node.

**Comment by AdoHe** [comment]
@zhouhaibing089 for your first question, unused image cleanup is already periodically, please refer [https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L947-L959](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L947-L959) for more details. Keeping unused images is necessary as we don't hope to always push the same image. For the ContainerGC implementation, please refer [https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockertools/container_gc.go#L182](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockertools/container_gc.go#L182) for more details. The `minimum-container-ttl-duration` is used to specify the ttl for dead containers.

**Comment by zhouhaibing089** [comment]
@AdoHe As you mean it is periodically, however it is a check on disk usage and decide whether to do garbage collection, what I mean is to do the real clean up on the unused image no matter how much the disk usage is.

Yes, for sure, avoiding pulling the same image is a good merit, but I am just thinking it **_may_** be better to cleanup it anyway, say once per day.

**Comment by AdoHe** [comment]
@zhouhaibing089 well, there is some tradeoff, just like what garbage collected language does, we reclaim garbage only when resource is not used anymore or in some limited environment, we do collect when resource usage exceed range.

**Comment by yujuhong** [comment]
> the threshold to delete image defaults to be 80%-90%, how about do this cleanup periodically? I am thinking that unused images on the node is not useful at all, the only situation it may help is that image could be reused when another scheduled pod uses it, too.

Kubelet now exposes images as part of the node status. Eventually, they will be incorporated into the scheduling decisions. Like @AdoHe said, we try to use the overall disk utilization as a signal to reclaim resources, but there are certainly some optimizations that we can do.

> maximum-dead-containers-per-container, maximum number of old instances to retain per container. Default is 2.
> I am confused here, it would be great help if you could help to explain it more here.

Assume there is a pod with two containers named FOO and BAR, and the restart policy is "Always". Every time when FOO or BAR exits, kubelet will create a new container to replace it. As time goes by, you now have 3 FOO exited containers and 5 BAR exited containers. Kubelet GC will look at the FOO and BAR containers separately, and ensure they each have no more than 2 exited containers. The reason behind keeping exited containers is that 1) kubelet needs to examine the exited container, and 2) kubelet relies on these tombstones as checkpoints (to know that it has started the container before). Both reasons can be tackled separately, but this is the status quo.

**Comment by zhouhaibing089** [comment]
@yujuhong I thought explanation saved my day, :)

If the images on that node are also reported as status to api-server and used by scheduler, then I would doubt on how much value it brings to us, maybe I have no insights here, two considerations come in my mind:

*   It may be the case that the same application would pick the same image, so it would have some side-effect to let the app get more possibility to be deployed on the same node.
*   Basically when the app upgrades, it generally come up with a new image tag(or using `latest` to make the image always get pulled), then the value it brings to us could only be some layer reusing.

**Comment by yujuhong** [comment]
Ah sorry..the image affinity scheduling PR has already been merged. See https://github.com/kubernetes/kubernetes/blob/master/docs/devel/scheduler_algorithm.md

> @yujuhong I thought your explanation saved my day, :)

You're welcome. I hope in the future we will be able to get rid of the flags that requires knowledge of kubelet implementation.

> It may be the case that the same application would pick the same image, so it would have some side-effect to let the app get more possibility to be deployed on the same node.

The scheduler relies on many priority functions, and one of them tries to spread pods with the same selector. I think the effect should be minimum, but @davidopp or @madhusudancs will know more.

> Basically when an app upgrades, it would mostly come up with a new image tag(or using latest to make the image always get pulled), then the value it brings to us could only be some layers reusing.

I don't think app upgrade is a good use case for image affinity. I think it's intended for applications that share some common containers.

**Comment by zhouhaibing089** [comment]
@yujuhong thanks, it does make sense.
